File: examples/basic_openai.py
"""
Example: Basic OpenAI integration with GenAI Telemetry
=====================================================

This example shows how to instrument a simple OpenAI chat application
with genai-telemetry for observability.

Prerequisites:
    pip install genai-telemetry openai

Usage:
    export OPENAI_API_KEY="your-api-key"
    python basic_openai.py
"""

from genai_telemetry import setup_telemetry, trace_llm

# Initialize telemetry with console output for development
setup_telemetry(
    workflow_name="openai-chatbot",
    exporter="console"  # Use "splunk", "elasticsearch", etc. in production
)


@trace_llm(model_name="gpt-4o-mini", model_provider="openai")
def chat(message: str) -> str:
    """
    Send a message to OpenAI and get a response.
    
    NOTE: We return the FULL response object to capture token counts,
    then extract the content for our return value.
    """
    from openai import OpenAI
    
    client = OpenAI()
    response = client.chat.completions.create(
        model="gpt-4o-mini",
        messages=[{"role": "user", "content": message}]
    )
    
    # Return full response - telemetry captures tokens
    return response


@trace_llm(model_name="gpt-4o-mini", model_provider="openai", extract_content=True)
def chat_simple(message: str) -> str:
    """
    Alternative: Use extract_content=True to automatically get just the text.
    """
    from openai import OpenAI
    
    client = OpenAI()
    response = client.chat.completions.create(
        model="gpt-4o-mini",
        messages=[{"role": "user", "content": message}]
    )
    
    return response  # Returns full response, but decorator extracts content


if __name__ == "__main__":
    print("=" * 60)
    print("Example 1: Full response (extract content manually)")
    print("=" * 60)
    
    response = chat("What is the capital of France?")
    print(f"\nResponse: {response.choices[0].message.content}")
    
    print("\n" + "=" * 60)
    print("Example 2: Auto-extracted content")
    print("=" * 60)
    
    content = chat_simple("What is 2 + 2?")
    print(f"\nResponse: {content}")
    
    print("\n" + "=" * 60)
    print("Check the console output above for telemetry data!")
    print("=" * 60)

File: js/examples/basic_openai.ts
/**
 * Basic OpenAI Example
 * 
 * This example demonstrates how to use genai-telemetry with OpenAI.
 * 
 * Prerequisites:
 *   npm install openai genai-telemetry
 *   export OPENAI_API_KEY=your-key
 */

import { setupTelemetry, traceLLM, traceEmbedding } from 'genai-telemetry';
import OpenAI from 'openai';

// Initialize telemetry with console output (for demo)
// In production, use Splunk, Elasticsearch, or another backend
setupTelemetry({
  workflowName: 'openai-example',
  exporter: 'console',
});

// Or use Splunk:
// setupTelemetry({
//   workflowName: 'openai-example',
//   exporter: 'splunk',
//   splunkUrl: 'http://splunk:8088',
//   splunkToken: 'your-token'
// });

const client = new OpenAI();

// Wrap the chat completion function
const chat = traceLLM(
  async (message: string) => {
    return await client.chat.completions.create({
      model: 'gpt-4o-mini',
      messages: [{ role: 'user', content: message }],
    });
  },
  { modelName: 'gpt-4o-mini', modelProvider: 'openai' }
);

// Wrap the embedding function
const embed = traceEmbedding(
  async (text: string) => {
    return await client.embeddings.create({
      model: 'text-embedding-3-small',
      input: text,
    });
  },
  { model: 'text-embedding-3-small' }
);

// Main function
async function main() {
  console.log('Starting OpenAI example...\n');

  // Chat completion
  console.log('1. Chat Completion:');
  const chatResponse = await chat('What is the capital of France?');
  console.log('Response:', chatResponse.choices[0].message.content);
  console.log('');

  // Embedding
  console.log('2. Embedding:');
  const embeddingResponse = await embed('Hello, world!');
  console.log('Embedding dimensions:', embeddingResponse.data[0].embedding.length);
  console.log('');

  // Multiple calls to show tracing
  console.log('3. Multiple calls:');
  const questions = [
    'What is 2+2?',
    'Who wrote Romeo and Juliet?',
    'What is the speed of light?',
  ];

  for (const question of questions) {
    const response = await chat(question);
    console.log(`Q: ${question}`);
    console.log(`A: ${response.choices[0].message.content}\n`);
  }

  console.log('Done! Check your telemetry backend for traces.');
}

main().catch(console.error);
